packages/inferno-router/__tests__/loaderOnRoute.spec.jsx

```diff
@@ -0,0 +1,586 @@
+    const [setDone, waitForRerender] = createEventGuard();
+    
+    const TEXT = "An error";
+    const loaderFunc = async () => {
+      setDone();
+      throw new Error(TEXT)
+    }
+
+    render(
+      <BrowserRouter initialEntries={['/']}>
+        <Route path="/" render={(props) => {
+          const err = useLoaderError(props);
+          return <h1>{err?.message}</h1>
+        }} loader={loaderFunc} />
+      </BrowserRouter>,
+      container
+    );
+
+    // Wait until async loader has completed
+    await waitForRerender();
+
+    expect(container.innerHTML).toContain(TEXT);
+  });
+
+  it('Can access initialData (for hydration)', async () => {
+    const TEXT = 'bubblegum';
+    const Component = (props, { router }) => {
+      const res = useLoaderData(props);
+      return <h1>{res?.message}</h1>
+    }
+
+    const loaderFunc = async () => { return { message: TEXT }}
+
+    const initialData = {
+      '/flowers': { res: await loaderFunc(), err: undefined, }
+    }
+
+    history.replaceState(undefined, undefined, '/flowers');
+    render(
+      <BrowserRouter initialData={initialData}>
+        <Route path="/flowers" render={Component} loader={loaderFunc} />
+      </BrowserRouter>,
+      container
+    );
+
+    expect(container.innerHTML).toContain(TEXT);
+  });
+
+  it('Should render component after after click', async () => {
+    const [setDone, waitForRerender] = createEventGuard();
+    const TEST = "ok";
+    const loaderFunc = async () => {
+      setDone();
+      return { message: TEST }
+    };
+
+    function RootComp() {
+      return <div id="root">ROOT</div>;
+    }
+
+    function CreateComp(props) {
+      const res = useLoaderData(props);
+      const err = useLoaderError(props);
+      return <div id="create">{res.message}</div>;
+    }
+
+    function PublishComp() {
+      return <div id="publish">PUBLISH</div>;
+    }
+
+    const tree = (
+      <BrowserRouter>
+        <div>
+          <nav>
+            <ul>
+              <li>
+                <NavLink exact to="/">
+                  Play
+                </NavLink>
+              </li>
+              <li id="createNav">
+                <NavLink to="/create">
+                  Create
+                </NavLink>
+              </li>
+              <li>
+                <NavLink to="/publish">
+                  Publish
+                </NavLink>
+              </li>
+            </ul>
+          </nav>
+          <Route exact path="/" component={RootComp} />
+          <Route path="/create" component={CreateComp} loader={loaderFunc} />
+          <Route path="/publish" component={PublishComp} />
+        </div>
+      </BrowserRouter>
+    );
+
+    render(tree, container);
+
+    expect(container.innerHTML).toContain("ROOT");
+
+    // Click create
+    const link = container.querySelector('#createNav');
+    link.firstChild.click();
+
+    // Wait until async loader has completed
+    await waitForRerender();
```

Comment:
inferno has a method called "rerender" which can be called to flush all pending renders, then there is no need to wait

Reply:
This is also waiting for the async loader to complete. I will look att combining the two.

Reply:
Fixed in 195f4686c (note: all the awaits are needed too)